### PR TITLE
Add default annual allowances table

### PR DESCRIPTION
**Pull request by samuelhwilliams**
This moves us from having default annual allowances for free text messages hard-coded to stored in the DB, which more closely follows what we do for text message rates.

Best to go commit-by-commit for this as there's some refactoring in the midst as well.

We also update the function to carry forward 0 allowances. This was previously handled by a special admin command, but it feels like this makes more sense to be default behaviour otherwise we still have to run that command ASAP at the start of each year to make sure zero allowances are carried forward.

---

Ticket: https://trello.com/c/jn2oqao9/335-move-the-default-annual-billing-allowances-from-code-to-the-db-similar-to-billing-rates

### CODE REVIEW

File: tests/app/dao/test_annual_billing_dao.py [hunk 1]
```diff
@@ -91,6 +90,17 @@ def test_set_default_free_allowance_for_service(notify_db_session, org_type, yea
     assert annual_billing[0].free_sms_fragment_limit == expected_default
 
 
+def test_set_default_free_allowance_for_service_fails_if_no_historical_default(notify_db_session):
+    service = create_service(organisation_type="central")
+
+    with pytest.raises(ValueError) as e:
+        set_default_free_allowance_for_service(service=service, year_start=2000)
```
Comment (CrystalPea) [inline tests/app/dao/test_annual_billing_dao.py:97]: Answering your commit message here:

Right now, if a user goes to `https://www.notifications.service.gov.uk/services/<service_name>/usage?year=2018` and their service existed then, they would see a valid result. 

This will break after those changes go in.
The user would have to manipulate their url to get there though, so maybe not worth pulling our hair out over :). But we could also avoid that by defaulting to 2020 allowance if year is older than 2020, since our allowances didn't change before 2020 as far as I'm aware.

**Review by CrystalPea — COMMENTED** [review]
Nice one, I left a few comments.

File: app/models.py [hunk 2]
```diff
@@ -611,6 +611,35 @@ def get_available_broadcast_providers(self):
...
+    Eg central government services get 40,000 free text messages for FY 2023.
+
+    The default rates will be applied to services automatically used when a new financial year begins. They can be
+    overridden on a per-service basis (eg some services may have their allowance reduced or removed).
+    """
+
+    __tablename__ = "default_annual_allowance"
+
+    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
+    valid_from_financial_year_start = db.Column(db.Integer, index=True, nullable=False)
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/models.py:610]: I did wonder if that was something we would do, but I don't think it actually makes any sense when diving down. Remember this is a default allowance, so this is only checked if they have no record in AnnualBilling yet. If we change the default halfway through the year, then it only applies for services who haven't interacted with services (either very stale ones or brand new ones). I'm not sure that we would ever change it mid-year anyway - we haven't needed to yet, so assuming YAGNI.

File: app/constants.py [hunk 3]
```diff
@@ -125,6 +125,30 @@
...
+    ORG_TYPE_CENTRAL,
+    ORG_TYPE_LOCAL,
+    ORG_TYPE_NHS_CENTRAL,
+    ORG_TYPE_NHS_LOCAL,
+    ORG_TYPE_NHS_GP,
+    ORG_TYPE_EMERGENCY_SERVICE,
+    ORG_TYPE_SCHOOL_OR_COLLEGE,
+    ORG_TYPE_OTHER,
+]
+CROWN_ORGANISATION_TYPES = ["nhs_central"]
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/constants.py:147]: Definitely does look weird on the surface. I honestly didn't give it much thought though, as it was just a copy/paste move.

File: tests/app/dao/test_annual_billing_dao.py (hunk 1 above)
Comment (samuelhwilliams) [reply tests/app/dao/test_annual_billing_dao.py:97]: > Right now, if a user goes to https://www.notifications.service.gov.uk/services/<service_name>/usage?year=2018 and their service existed then, they would see a valid result.

I don't think it _should_ break if the service did already exist in 2018, as they should have an entry in AnnualBilling already, which means they won't need to query the DefaultAnnualAllowance table.

However, I think you make a point that if a service which _didn't_ exist in 2018 hit the same URL, they'll currently see a valid page whereas after this change they'll get a 500.

I'll fix that.

Comment (samuelhwilliams) [reply tests/app/dao/test_annual_billing_dao.py]: Fixed in 3843d44dad0c3a2ba973498a0c1a5780ff7a3702 (which I'll squash before merging)

Comment (samuelhwilliams) [reply tests/app/dao/test_annual_billing_dao.py]: Also made me spot an edge case that would let users 'lock in' allowances for future years, which I've closed in 0046c5d53ff9cb68e06dfdf782eae52737d663e0

File: app/constants.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/constants.py:147]: This is the PR and commit that added it: https://github.com/alphagov/notifications-api/pull/2557, d89ef0594f6318fc64a569a4d1f736d1d98848d3.

It doesn't make it super clear _why_ `nhs_central` is a crown org. But it does highlight another weirdness, I think, which is that we store this info both as constants and in the DB in the `organisation_types` table. We probably should clean this up and only reference the DB values. 🤔

File: migrations/versions/0409_annual_allowance.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    #         valid_from_financial_year_start=valid_from_financial_year_start,
+    #         organisation_type=org_type,
+    #         allowance=allowance,
+    #         notification_type="sms",
+    #     )
+    #     for org_type, allowances in ORIGINAL_ALLOWANCES.items()
+    #     for valid_from_financial_year_start, allowance in allowances.items()
+    # ]
+    # ---
+    # See above code fragment used to generate the below dict
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply migrations/versions/0409_annual_allowance.py:108]: Makes sense, thank you!

File: app/constants.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (CrystalPea) [reply app/constants.py:147]: Ok, do you prefer to do it in this PR, or would it be better to add a card to the longer backlog about this?

[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/constants.py:147]: I think one to do separately as it's broadly unrelated and has been long standing behaviour.

**Review by CrystalPea — APPROVED** [review]
